### PR TITLE
handle upgrade from 2.2.X to 2.3 to remove finalizer in clusterrole

### PR DESCRIPTION
**Pull request by ldpliu**
https://github.com/open-cluster-management/backlog/issues/13670

### CODE REVIEW

**Comment by zhiweiyin318** [comment]
I don't think it is a good option to create a controller to cleanup the finalizer only for the upgrade case.
could you consider to finish the cleanup job once at the beginning or cleanup the finalizer when the clusterrole is reconciled.
cc @qiujian16

**Comment by ldpliu** [comment]
> I don't think it is a good option to create a controller to cleanup the finalizer only for the upgrade case.
> could you consider to finish the cleanup job once at the beginning or cleanup the finalizer when the clusterrole is reconciled.
> cc @qiujian16

Currentlly, it's not a real controller, and it's a go-rotine and will only start on controller start and will finish after the finalizer cleaned.

File: pkg/controllers/gc/garbagecollector_controller.go [hunk 1]
```diff
@@ -81,3 +87,47 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
+	return &CleanGarbageFinalizer{
+		kubeClient: kubeClient,
+		ctx:        ctx,
+		stopCh:     stopCh,
+	}
+}
+
+// start a routine to sync the clusterrolebinding periodically.
+func (c *CleanGarbageFinalizer) Run() {
+	go utilwait.Until(c.clean, 60*time.Second, c.stopCh)
```
Comment (qiujian16) [inline pkg/controllers/gc/garbagecollector_controller.go:109]: if you run go outside, you do not need goroutine here. And you can use wait.PollImmediateUntil so no need to close channel in clean.

File: pkg/controllers/gc/garbagecollector_controller.go [hunk 2]
```diff
@@ -81,3 +87,47 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
+	if err != nil {
+		return
+	}
+	allFinalizerRmoved := true
+	for _, clusterrole := range clusterRolebindingList.Items {
+		if !utils.ContainsString(clusterrole.GetFinalizers(), ClustersetFinalizerName) {
+			continue
+		}
+
+		klog.Infof("removing ClusterRoleBinding Finalizer in ClusterRole %v", clusterrole.Name)
```
Comment (qiujian16) [inline pkg/controllers/gc/garbagecollector_controller.go:123]: V(4) please

**Review by qiujian16 — COMMENTED** [review]
/approve
/lgtm

**Pull request merged**